frontend/src/features/admin-form/create/logic/components/LogicContent/EditCondition/ThenShowBlock.tsx

```diff
@@ -0,0 +1,229 @@
+    // Default to `show`
+    return logicTypeValue === LogicType.PreventSubmit
+      ? 'preventSubmitMessage'
+      : 'show'
+  }, [logicTypeValue])
+
+  return (
+    <Stack
+      direction="column"
+      spacing="0.75rem"
+      py="1.5rem"
+      px={{ base: '1.5rem', md: '2rem' }}
+    >
+      <Stack
+        direction={{ base: 'column', md: 'row' }}
+        spacing={{ base: 0, md: '0.5rem' }}
+      >
+        <BlockLabelText id="logicType-label" htmlFor="logicType">
+          Then
+        </BlockLabelText>
+        <FormControl
+          isReadOnly={isLoading}
+          id="logicType"
+          isRequired
+          isInvalid={!!errors.logicType}
+        >
+          <Controller
+            name="logicType"
+            control={control}
+            rules={{
+              required: 'Please select logic type.',
+            }}
+            render={({ field }) => (
+              <SingleSelect
+                isDisabled={isLoading}
+                isSearchable={false}
+                isClearable={false}
+                placeholder="Select a type of result"
+                items={logicTypeItems}
+                {...field}
+              />
+            )}
+          />
+          <FormErrorMessage>{errors.logicType?.message}</FormErrorMessage>
+        </FormControl>
+      </Stack>
+
+      <Stack
+        direction={{ base: 'column', md: 'row' }}
+        spacing={{ base: 0, md: '0.5rem' }}
+      >
+        <BlockLabelText
+          id={`${currentShowLabel}-label`}
+          htmlFor={currentShowLabel}
+        >
+          Show
+        </BlockLabelText>
+        <ThenLogicInput
+          formFields={formFields}
+          mapIdToField={mapIdToField}
+          formMethods={formMethods}
+          isLoading={isLoading}
+        />
+      </Stack>
+    </Stack>
+  )
+}
+
+const ThenLogicInput = ({
+  isLoading,
+  formMethods,
+  formFields,
+  mapIdToField,
+}: ThenShowBlockProps) => {
+  const {
+    watch,
+    control,
+    register,
+    getValues,
+    formState: { errors },
+  } = formMethods
+
+  const logicTypeValue = watch('logicType')
+  const logicConditionsWatch = useWatchDependency(watch, 'conditions')
+
+  const thenValueItems = useMemo(() => {
+    // Return every field except fields that are already used in the logic.
+    if (logicTypeValue === LogicType.ShowFields) {
+      const usedFieldIds = new Set(
+        logicConditionsWatch.value.map((condition) => condition.field),
+      )
+      if (!formFields || !mapIdToField) return []
+      return formFields
+        .filter((f) => !usedFieldIds.has(f._id))
+        .map((f) => {
+          const mappedField = mapIdToField[f._id]
+          return {
+            value: f._id,
+            label: `${mappedField.questionNumber}. ${mappedField.title}`,
+            icon: BASICFIELD_TO_DRAWER_META[f.fieldType].icon,
+          }
+        })
+    }
+    return []
+    // Watch entire <***>Watch variables since <***>Watch.value is a proxy
```

Comment:
what do you mean by "proxy"? sorry don't really understand this comment - do you mean it's a reference to an object?

Reply:
Ah, [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) is a Javascript object (and a design pattern). `watch` uses `Proxy`s under the hood which means value changes may not result in rerenders (which is not what we want).

Read more: https://react-hook-form.com/api/useform/watch#:~:text=watch%20result%20is%20optimised%20for%20render%20phase%20instead%20of%20useEffect%27s%20deps%2C%20to%20detect%20value%20update%20you%20may%20want%20to%20use%20an%20external%20custom%20hook%20for%20value%20comparison.